packages/dds/merge-tree/Obliterate.md

```diff
@@ -0,0 +1,408 @@
+either the obliterated region should expand to include that text, or it should not.
+This topic will be covered in the [endpoint behavior](#endpoint-behavior) section,
+but for implementation strategy discussion one should assume that the implementation should support both options (and leave it up to applications
+to specify).
+
+## Implementation Strategy
+
+This section is focused on implementation of the "obliterate" semantics inside merge tree.
+This will constitute the bulk of the complexity of the feature.
+Since obliterate is generally a "different kind of remove," there may be a nice abstraction to introduce at the code level to generalize
+removal information. However, in favor of introducing niceties later this design document will assume fields are inlined and focus on
+the strategy for ensuring eventual consistency. If such an abstraction is introduced, ideally it would enable better "pay-to-play" of
+common code paths based on merge-tree feature usage.
+As an example, `BaseSegment.split()` needs to copy segment properties to the split segment.
+So new properties added to segment will unnecessarily copy undefined values.
+
+There are a few aspects of merge tree's bookkeeping and general feature set that require consideration when designing new op semantics:
+
+- Any changes to direct fields of tree nodes themselves (either new data or changes to bookkeeping of existing data)
+- How the feature interacts with an increasing collab window and zamboni
+- Impact on the partial lengths scheme
+- Bookkeeping and handling of overlapping removals (note some may be obliterates and some may not be)
+- Reconnection
+- Snapshotting impact
+
+We'll first present an overview of a potential scheme for implementing the obliterate op, then comment on these aspects.
+
+### High-level bookkeeping changes
+
+Segments will be augmented with `movedSeq` and `localMovedSeq` fields which generally align with the semantics of `seq, localSeq, removedSeq,` and `localRemovedSeq`.
+When segments are moved and not just obliterated, they will also contain a reference to the destination segment.
+This may look as follows:
+
+```typescript
+/**
+ * Tracks information about when and where this segment was moved to.
+ * @example - Suppose a merge tree had 3 TextSegments "X", "A", and "B", and
+ * received the operation `move({ start: 0, end: 1 }, { dest: 3 }, { seq: 30 })` (moving the "X"
+ * after the "A" and the "B").
+ * After processing this operation, it would have the segments `[<moved "X" tombstone>, "A", "B", "X"]`.
+ * The moved "X" tombstone segment would have the following IMoveInfo: `{ movedSeq: 30, moveDst: <reference to living "X" segment>}`
+ */
+export interface IMoveInfo {
+    /**
+     * Local seq at which this segment was moved if the move is yet-to-be acked. Only set on the tombstone "source" segment of the move.
+     */
+    localMovedSeq?: number;
```

Comment:
I was imagining move would put `seq` on the move destination equal to the moved seq. This will make lengths calculations generally right (ignoring overlapping cases). The only negative impact of it is how it interplays with attribution. But I may reconsider after thinking about that a bit more.

Reply:
actually, we need `movedClientId?: number` on the segment as well. will update once I'm done thinking through how length tracking works.

Reply:
this is updated. I think we should just make semantics of `clientId` be the client that caused the segment to be inserted/moved, and if we want separately in the annotation design we can track that elsewhere (could even use props or some other feature, though we'd want special handling for compression)